core/src/main/java/io/questdb/griffin/DatabaseSnapshotAgent.java

```diff
@@ -0,0 +1,297 @@
+        final CharSequence root = configuration.getRoot();
+        final CharSequence snapshotRoot = configuration.getSnapshotRoot();
+
+        try (Path path = new Path(); Path copyPath = new Path()) {
+            path.of(snapshotRoot);
+            final int snapshotRootLen = path.length();
+            copyPath.of(root);
+            final int rootLen = copyPath.length();
+
+            // Check if the snapshot dir exists.
+            if (!ff.exists(path.slash$())) {
+                return;
+            }
+
+            // Check if the snapshot metadata file exists.
+            path.trimTo(snapshotRootLen).concat(TableUtils.SNAPSHOT_META_FILE_NAME).$();
+            if (!ff.exists(path)) {
+                return;
+            }
+
+            // Check if the snapshot instance id is different from what's in the snapshot.
+            try (MemoryCMARW mem = Vm.getCMARWInstance()) {
+                mem.smallFile(ff, path, MemoryTag.MMAP_DEFAULT);
+
+                final CharSequence currentInstanceId = configuration.getSnapshotInstanceId();
+                final CharSequence snapshotInstanceId = mem.getStr(0);
+                if (Chars.equals(currentInstanceId, snapshotInstanceId)) {
+                    return;
+                }
+
+                LOG.info()
+                        .$("starting snapshot recovery [currentId=`").$(currentInstanceId)
+                        .$("`, previousId=`").$(snapshotInstanceId)
+                        .$("`]").$();
+            }
+
+            // OK, we need to recover from the snapshot.
+            AtomicInteger recoveredMetaFiles = new AtomicInteger();
+            AtomicInteger recoveredTxnFiles = new AtomicInteger();
+            AtomicInteger recoveredCVFiles = new AtomicInteger();
+            path.trimTo(snapshotRootLen).concat(configuration.getDbDirectory()).$();
+            final int snapshotDbLen = path.length();
+            ff.iterateDir(path, (pUtf8NameZ, type) -> {
+                if (Files.isDir(pUtf8NameZ, type)) {
+                    path.trimTo(snapshotDbLen).concat(pUtf8NameZ);
+                    copyPath.trimTo(rootLen).concat(pUtf8NameZ);
+                    final int plen = path.length();
+                    final int cplen = copyPath.length();
+
+                    path.concat(TableUtils.META_FILE_NAME).$();
+                    copyPath.concat(TableUtils.META_FILE_NAME).$();
+                    if (ff.exists(path) && ff.exists(copyPath)) {
+                        if (ff.copy(path, copyPath) < 0) {
+                            LOG.error()
+                                    .$("could not copy snapshot _meta file [src=").$(path)
+                                    .$(", dst=").$(copyPath)
+                                    .$(", errno=").$(ff.errno())
+                                    .$(']').$();
+                        } else {
+                            recoveredMetaFiles.incrementAndGet();
```

Comment:
Would be good to log success recoveries here too in case it's aborted half way

Reply:
Added intermediate logging in 3e2f0b4cea1601b70e737b15316c1f8e85b8a3b9

It might be too chatty if the number of tables is high, but probably even then it should be more or less OK.